tsdb/querier.go

```diff
@@ -387,39 +387,59 @@ func postingsForMatcher(ix IndexReader, m *labels.Matcher) (index.Postings, erro
 		}
 	}
 
-	vals, err := ix.LabelValues(m.Name)
+	vals, err := ix.LabelValuesUnsorted(m.Name)
 	if err != nil {
 		return nil, err
 	}
 
 	var res []string
+	lastVal, isUnsorted := "", false
 	for _, val := range vals {
 		if m.Matches(val) {
 			res = append(res, val)
+			if val < lastVal {
+				isUnsorted = true
+			}
+			lastVal = val
 		}
 	}
 
 	if len(res) == 0 {
 		return index.EmptyPostings(), nil
 	}
 
+	if isUnsorted {
+		// Label values coming from blocks are always sorted.
+		// So we sort here only if it was actually unsorted.
+		sort.Strings(res)
+	}
 	return ix.Postings(m.Name, res...)
 }
 
 // inversePostingsForMatcher returns the postings for the series with the label name set but not matching the matcher.
 func inversePostingsForMatcher(ix IndexReader, m *labels.Matcher) (index.Postings, error) {
-	vals, err := ix.LabelValues(m.Name)
+	vals, err := ix.LabelValuesUnsorted(m.Name)
 	if err != nil {
 		return nil, err
 	}
 
 	var res []string
+	lastVal, isUnsorted := "", false
 	for _, val := range vals {
 		if !m.Matches(val) {
 			res = append(res, val)
+			if val < lastVal {
```

Comment:
Do all these comparisons cost us more than sorting an already-sorted list?

Reply:
Probably not :P I will revert this